### DISCUSSION TITLE
D3DCompiler fails to compile shader and terminates with error X3000

### DISCUSSION THREAD

**Discussion by itsspike** [dotnet/Silk.NET#824] [Q&A]
# Summary
I've tried to compile a very basic shader with the D3DCompiler and it's Compile function and CompileFromFile function respectively. However, the compiling fails. The `errorMsgs` variable says in both cases `error X3000`. Please see the MWE and detailed error messages below.

# Steps to reproduce
- Platform: Desktop
- Framework Version: .NET 6
- API: DirectX
- API Version: DirectX12

## Using D3DCompiler.Compile()

```csharp
D3DCompiler D3D = D3DCompiler.GetApi();

var filePath = Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location)!, "HelloTriangle.hlsl");
if (!File.Exists(filePath))
{
	throw new Exception("huh?");
}

string shaderContent = File.ReadAllText(filePath);

ID3D10Blob* shader;
ID3D10Blob* errorMsgs;

fixed (char* pContent = shaderContent)
{
	int res = D3D.Compile(pSrcData: pContent,
	                      SrcDataSize: (nuint)shaderContent.Length,
	                      pSourceName: "triangle",
	                      pDefines: null,
	                      pInclude: null,
	                      pEntrypoint: "VSMain",
	                      pTarget: "vs_5_0",
	                      Flags1: 0,
	                      Flags2: 0,
	                      ppCode: &shader,
	                      ppErrorMsgs: &errorMsgs);

	SilkMarshal.ThrowHResult(res);
}
```

If I look into the error message, I get: `...\Debug\net6.0\triangle(1,1): error X3000: unrecognized identifier 's'`

## Using D3DCompiler.CompileFromFile()

```csharp
D3DCompiler D3D = D3DCompiler.GetApi();

var filePath = Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location)!, "HelloTriangle.hlsl");

ID3D10Blob* shader;
ID3D10Blob* errorMsgs;

fixed (char* pPath = filePath)
{
	int res = D3D.CompileFromFile(pFileName: pPath, 
	                              pDefines: null, 
	                              pInclude: null, 
	                              pEntrypoint: "VSMain", 
	                              pTarget: "vs_5_0", 
	                              Flags1: 0, 
	                              Flags2: 0, 
	                              ppCode: &shader, 
	                              ppErrorMsgs: &errorMsgs);

	SilkMarshal.ThrowHResult(res);
}
```

If I look into the error message, I get: `...\Debug\net6.0\HelloTriangle.hlsl(1,1): error X3000: Illegal character in shader file`

The shader I've tried to use: https://github.com/dotnet/Silk.NET/blob/e79e16542e0e916be8aa8460fa06eed2719d8430/src/Lab/Experiments/D3D12Triangle/HelloTriangle.hlsl

**Comment by abenedik** [comment]
Check the hlsl file encoding. It should be ANSI. If it is "UTF with signature" then the file starts with two bytes that represent the UTF signature and this produces the "error X3000: Illegal character in shader file" error.

You can change the encoding in Visual Studio by opening a file and then selecting "Save [FileName] As" menu option. Then in the dialog click on the arrow in the Save button and select "Save with Encoding..."

**Comment by Perksey** [comment]
+1 probably has a BOM

**Comment by itsspike** [comment]
Thank you, that helped! I'm now able to compile the shader with the `CompileFromFile` function. But the error with `Compile` remains the same (unrecognized identifier 's'). Am I missing something?
